File: exercises/kth_smallest_BST_230.py
class Solution:
    count: int = 0
    result: int | None = None
    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        self.solve(root, k)
        return self.result

    def solve(self, node: TreeNode | None, k: int):
        if not node: return
        self.solve(node.left, k)
        if self.result is not None: return
        self.count += 1
        if self.count == k: self.result = node.val
        self.solve(node.right, k)

File: exercises/kth_smallest_BST_230.ts
// Rec Solution with nested function
function kthSmallest(root: TreeNode | null, k: number): number {
    let count = 0; 
    let result: number | null = null; 
    const traverse = function(node: TreeNode | null) {
        if (node === null || result !== null) return; 
        traverse(node.left);
        count++;
        if (count === k) {
            result = node.val; 
            return; 
        }
        traverse(node.right);
    };
    traverse(root);
    return result!;
};
// Iter Solution
//function kthSmallest(root: TreeNode | null, k: number): number {
//	if (!root) return -1;
//	const stack: TreeNode[] = [];
//	while (root) {
//		stack.push(root);
//		root = root.left;
//	}
//	while (k !== 0) {
//		const node = stack.pop()!;
//		k--;
//		if (k === 0) return node.val;
//		let right = node.right;
//		while (right) {
//			stack.push(right);
//			right = right.left;
//		}
//	}
//	return -1;
//}
